### PR TITLE
Different fixes (MLD, LE-systems, valgrind, etc)

### PR DESCRIPTION
**Pull request by ghost**
Greetings. Thanks for good project (especially for MLD-proxy implementation).

For more information see commit messages and code.

Also there were some troubles with commits 4a7f5e2c4c11 9c5c26ca and imp_get_mfcc_ttls function (e.g. router interface with ifindex > MAXVIFS doesn't work). I thought this project is dead. So I simply reverted this commits, removed  ifindex checking  and forked your repo.
But if it's still alive, I could provide more patches and more tests results (ofc if you are interested in this).

Thanks.
BR Konstantin

### CODE REVIEW

**Comment by haibbo** [comment]
@zentorin 
Yes, "improxy" is alive. I really appreciate your great work, and welcome your coming patches.
I will take some time to review these commits for merging them.

File: src/proxy.c [hunk 1]
```diff
@@ -362,7 +362,10 @@ STATUS init_mproxy4(mcast_proxy *p_mp)
         ui = 1;
         if (setsockopt(p_mp->igmp_socket, IPPROTO_IP, IP_PKTINFO, &ui, sizeof(ui)))
             IMP_LOG_ERROR("IP_PKTINFO: %s\n", strerror(errno));
-
+        ui = 0xC0; /* set DSCP mark CS6 for outgoing */
```
Comment (haibbo) [inline src/proxy.c]: @zentorin  Why setting DSCP mark to CS6?

**Comment by zentorin** [comment]
It's optional.
Some programs use this value by default for outgoing IGMP packets.
But I checked again different RFCs and found only this:
https://tools.ietf.org/html/rfc4594
> If IGMP is used for multicast session control such as channel
>       changing in IPTV systems, then IGMP packets should be marked with
>       CS5 DSCP (use Signaling service class).  When IGMP is used only
>       for the normal multicast routing purpose, it should be marked with
>       CS6 DSCP (use Network Control service class).

So, you may remove this code or add something like vlc's --dscp option.

File: src/proxy.c (hunk 1 above)
Comment (zentorin) [reply src/proxy.c]: It's optional.
Some programs use this value by default for outgoing IGMP packets.
But I checked again different RFCs and found only this:
https://tools.ietf.org/html/rfc4594
> If IGMP is used for multicast session control such as channel
>       changing in IPTV systems, then IGMP packets should be marked with
>       CS5 DSCP (use Signaling service class).  When IGMP is used only
>       for the normal multicast routing purpose, it should be marked with
>       CS6 DSCP (use Network Control service class).

So, you may remove this code or add something like vlc's --dscp option.

File: src/output.c [hunk 2]
```diff
@@ -275,15 +268,33 @@ void send_igmp_mld_query(imp_interface *p_if, im_version version,
...
+        struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+        cmsg->cmsg_level = IPPROTO_IPV6;
+        cmsg->cmsg_type = IPV6_PKTINFO;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(struct in6_pktinfo));
+
+        struct in6_pktinfo *pktinfo = (struct in6_pktinfo *)CMSG_DATA(cmsg);
+        pktinfo->ipi6_ifindex = p_if->if_index;
+        memcpy(&pktinfo->ipi6_addr, &p_if->if_addr.v6.sin6_addr, sizeof(struct in6_addr));
+
+        if (sendmsg(socket, &msg, 0) < 0) {
```
Comment (haibbo) [inline src/output.c]: @zentorin 
In my test, MLD query can be sent with a valid IPv6 link-local source address using 'sendto'.
Can you provide more details about replacing it with sendmsg.

File: src/proxy.c (hunk 1 above)
Comment (haibbo) [reply src/proxy.c]: Thanks, i will keep it.

File: src/output.c (hunk 2 above)
Comment (zentorin) [reply src/output.c]: For sendto it depends on query's destination address.
If it has link-local scope, then query will be sent with link-local src address.
If it has global scope, then query will be sent with global address.

General queries have ff02::1 destination address, so they have correct src address even with sendto. But Address Specific Queries may have global address (e.g. ff0e::1) as a destination address. As a result, they will have global src address. Clients will drop such queries.

**Pull request merged**